### PR TITLE
Numerical absorption refactor

### PR DESCRIPTION
**Pull request by peterfpeterson**
This PR makes way for `Material` to calculate the linear absorption coefficients, and changes the integration technique to use to use the more stable [pairwise summation](https://en.wikipedia.org/wiki/Pairwise_summation). The previous method of summing the elements gave different results when changing the order of the elements.

Unfortunately, further refactoring of the code to make `doIntegration` to be free functions made the code less clear. The stability of the current method was verified by changing the use on L210 for diffraction to be
```C++
if (L2s.size() > MAX_INTEGRATION_LENGTH) {
  Y[j] = this->doIntegration(-linearCoefAbs[j], L2s, MAX_INTEGRATION_LENGTH, L2s.size()) +
         this->doIntegration(-linearCoefAbs[j], L2s, 0, MAX_INTEGRATION_LENGTH);
} else {
  Y[j] = this->doIntegration(-linearCoefAbs[j], L2s, 0, L2s.size());
}
```
which gave the same results for the new unit test as the proposed version. The selection of the parameter, `MAX_INTEGRATION_LENGTH`, is (mostly) arbitrarily, but is significantly smaller than the 3300 elements that the WISH system test uses.

**To test:**

Short of doing a full study of the numerical stability of the new method, I don't have a good recommendation on how to properly test this. Since the existing unit tests weren't changed the refactor did not harm existing calculations. It is only the cases of decreasing the rasterization element sizes that will expose the issue.

*There is no associated issue.*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/AbsorptionCorrection.cpp [hunk 1]
```diff
@@ -35,11 +35,31 @@ using namespace Kernel;
...
+      static_cast<size_t>(floor(.5 * (static_cast<double>(stop - start))));
+  return start + half;
+}
+
+double energyToWavelength(const double energyFixed) {
+  Unit_const_sptr energy = UnitFactory::Instance().create("Energy");
+  double factor, power;
+  energy->quickConversion(*UnitFactory::Instance().create("Wavelength"), factor,
+                          power);
+  return factor * std::pow(energyFixed, power);
```
Comment (soininen) [inline Framework/Algorithms/src/AbsorptionCorrection.cpp]: There is a `E_mev_toNeutronWavenumberSq` in `PhysicalConstants.h`. Could this function be shortened to
```c++
return std::sqrt(energyFixed / E_mev_toNeutronWavenumberSq);
```
?

File: Framework/Algorithms/src/AbsorptionCorrection.cpp [hunk 2]
```diff
@@ -163,42 +183,42 @@ void AbsorptionCorrection::exec() {
...
-      if (m_emode == 0) // Elastic
-      {
-        Y[j] = this->doIntegration(lambda, L2s);
-      } else if (m_emode == 1) // Direct
-      {
-        Y[j] = this->doIntegration(m_lambdaFixed, lambda, L2s);
-      } else if (m_emode == 2) // Indirect
-      {
-        Y[j] = this->doIntegration(lambda, lambda_f, L2s);
+      if (m_emode == 0) { // Elastic
```
Comment (soininen) [inline Framework/Algorithms/src/AbsorptionCorrection.cpp]: I think we have `enum`s for these, please check `DeltaEMode.h`.

File: Framework/Kernel/src/Material.cpp [hunk 3]
```diff
@@ -243,6 +243,31 @@ double Material::absorbXSection(const double lambda) const {
...
+std::vector<double> Material::linearAbsorpCoef(
+    std::vector<double>::const_iterator lambdaBegin,
+    std::vector<double>::const_iterator lambdaEnd) const {
+  const double linearCoefByWL =
+      absorbXSection(PhysicalConstants::NeutronAtom::ReferenceLambda) * 100. *
+      numberDensity() / PhysicalConstants::NeutronAtom::ReferenceLambda;
+
+  std::vector<double> linearCoef(std::distance(lambdaBegin, lambdaEnd));
+
+  std::transform(lambdaBegin, lambdaEnd, linearCoef.begin(),
```
Comment (soininen) [inline Framework/Kernel/src/Material.cpp]: :+1: for `std::transform`.

**Review by soininen — CHANGES_REQUESTED** [review]
The refactored code looks good, just a few comments below on some old code carried over while copy-pasting. A unit test has been added as well as release notes so otherwise this looks good to go.

File: Framework/Algorithms/src/AbsorptionCorrection.cpp (hunk 2 above)
Comment (soininen) [reply Framework/Algorithms/src/AbsorptionCorrection.cpp]: Maybe even try out a `switch` statement here?

File: Framework/Algorithms/src/AbsorptionCorrection.cpp (hunk 1 above)
Comment (peterfpeterson) [reply Framework/Algorithms/src/AbsorptionCorrection.cpp]: Good point.

File: Framework/Algorithms/src/AbsorptionCorrection.cpp (hunk 2 above)
Comment (peterfpeterson) [reply Framework/Algorithms/src/AbsorptionCorrection.cpp]: I was bothered by the `int`. Thanks for letting me know about the `enum`.

**Review by soininen — APPROVED** [review]

**Pull request merged**